Fix create-new-model for for Controllers

hatched opened the pull request:
Now that the only place you can create a model is in the user profile view this adds the ability to specify the model name at time of creation and then connects you to that model.

hatched commented:
#### To QA

jujugui commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5190/
Test FAILed.

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5191/
Test PASSed.

huwshimi commented on jujugui/static/gui/src/app/components/user-profile/_user-profile.scss:
```diff
@@ -1,7 +1,67 @@
+            white-space: nowrap;
+            transition: all 0.5s ease-in;
+            width: 150px;
+            outline: none;
```
You don't need this, I just had to add that as part of the demo.

hatched replied:
I actually do, without it it puts that blue ring around the input which makes the animation look less attractive :)

huwshimi replied:
That's strange, it should be covered by this: https://github.com/juju/juju-gui/blob/develop/jujugui/static/gui/src/app/assets/css/_generic.scss#L9

huwshimi replied:
When I just tried removing it it worked fine.

hatched replied:
Alright, removed. Thanks.

bac commented:
QA-almost

I did everything you requested and it all worked.  Created lots of environments and it behaved exactly as expected.

Created 'env1' just fine.

But when I went back to the profile and tried to create 'env1' again, the gui attempted to open but failed.  Error console showed 'ReferenceError: err is not defined'

huwshimi commented on jujugui/static/gui/src/app/components/user-profile/_user-profile.scss:
```diff
@@ -1,7 +1,67 @@
+            transition: all 0.5s ease-in;
+            width: 150px;
+            outline: none;
+            padding: 10px 0;
```
You'll probably don't want to set up the top and bottom padding here, if you need to remove the side padding do `padding-right: 0; padding-left: 0;`

hatched replied:
I removed it entirely as it wasn't needed.

hatched replied:
turns out I was wrong, it is needed and I've implemented it as you suggested.

huwshimi commented on jujugui/static/gui/src/app/components/user-profile/_user-profile.scss:
```diff
@@ -1,7 +1,67 @@
+
+            &.error {
+                .deployment-input__field {
+                    border: 1px solid #df382c;
```
You could just do `border: 1px solid $error;`

hatched replied:
done

huwshimi commented on jujugui/static/gui/src/app/components/user-profile/_user-profile.scss:
```diff
@@ -1,7 +1,67 @@
+
+        .deployment-input__errors {
+            position: absolute;
+            z-index: 600;
```
Could you please use a z-index from our list: https://github.com/juju/juju-gui/blob/develop/jujugui/static/gui/src/app/assets/css/base.scss#L34

e.g. https://github.com/juju/juju-gui/blob/develop/jujugui/static/gui/src/app/assets/css/_header.scss#L13

hatched replied:
Thanks for the example usage. Done.

huwshimi commented on jujugui/static/gui/src/app/components/user-profile/user-profile.js:
```diff
@@ -209,6 +211,39 @@ YUI.add('user-profile', function() {
+        data => {
+          if (data.err) {
+            // XXX Throw notification
+            console.log(err);
```
`console.error`

hatched replied:
done, although this is also going to have real error handling implemented to resolve the qa issue @bac found.

huwshimi commented on jujugui/static/gui/src/app/components/user-profile/user-profile.js:
```diff
@@ -209,6 +211,39 @@ YUI.add('user-profile', function() {
+      }
+      // Because this will automatically connect to the model lets show
+      // the connecting mask right now.
+      this.props.showConnectingMask();
```
Do we show this now because it takes some time for the model to get created before it is switched to (`switchModel` below shows the mask too)?

hatched replied:
That's right - on my connection (which is quite poor) this takes a few seconds to loop around

huwshimi commented on jujugui/static/gui/src/app/components/user-profile/user-profile.js:
```diff
@@ -371,6 +406,58 @@ YUI.add('user-profile', function() {
+              }]} />
+            <juju.components.GenericButton
+              action={this.createAndSwitch}
+              type='inline-neutral second'
```
Double quotes with JSX please.

hatched replied:
done

huwshimi commented on jujugui/static/gui/src/app/components/user-profile/user-profile.js:
```diff
@@ -371,6 +406,58 @@ YUI.add('user-profile', function() {
+            <juju.components.GenericButton
+              action={this.createAndSwitch}
+              type='inline-neutral second'
+              title='Submit' />
```
Double quotes with JSX please.

hatched replied:
done

huwshimi commented:
I could have missed it, but I couldn't see any code that only displays this form when using a controller.

A couple of QA things:
- if your profile is empty the button appears to the right instead of centred and the errors appear up above the form next to the image
- once you've click 'Create new' there is no way to cancel

huwshimi commented on jujugui/static/gui/src/app/components/user-profile/user-profile.js:
```diff
@@ -209,6 +211,39 @@ YUI.add('user-profile', function() {
+      // Because this will automatically connect to the model lets show
+      // the connecting mask right now.
+      this.props.showConnectingMask();
+      // XXX This is only for the JIMM flow.
```
I thought we wanted this only for use with a controller, is a controller a JIMM now?

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5195/
Test PASSed.

hatched commented:
Good catch @bac I have added the error handling to surface the error to the user.

Thanks for the comments @huwshimi - I've fixed them all but left out the ability to cancel a model creation. 

I've updated the qa instructions for the new features added.

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5196/
Test PASSed.

huwshimi commented on jujugui/static/gui/src/app/components/deployment/input/_input.scss:
```diff
@@ -12,8 +12,6 @@
         height: 20px;
         margin-bottom: 20px;
-        padding-top: 25px;
-        padding-bottom: 7px;
```
You'll need to add the styles somewhere else or apply a `--without-label` class that removes these styles or something as currently with this change the inputs in the deployment flow are broken.

hatched replied:
I actually overwrite the padding entirely in the _user-profile.scss so I've reverted this change. Good catch thx

huwshimi commented on jujugui/static/gui/src/app/components/user-profile/user-profile.js:
```diff
@@ -209,6 +213,45 @@ YUI.add('user-profile', function() {
+      @method createAndSwitch
+      @param {Object} e The event handler from a form submission.
+    */
+    createAndSwitch: function(e) {
```
I feel like this might be better off as a util so that it can be used by other components, however, if this is going to all be removed shortly then we could leave it here. As a util it would also mean you don't have to pass as many props to this component.

hatched replied:
Yeah I was thinking the same but hopefully all of this will go away soon. (famous last words ;) )

huwshimi commented:
QA OK :+1: with a few little comments.

When I QAed and added a model and viewed the profile it showed all the models as "is being destroyed" and the model names did not appear in the env switcher, but I assume that is all because of our controller set up or something.

anthonydillon commented on jujugui/static/gui/src/app/components/user-profile/_user-profile.scss:
```diff
@@ -1,7 +1,67 @@
+        &.collapsed {
+            .deployment-input__field,
+            .button--inline-neutral.second {
+                width: 0;
```
Could you use display: none?

hatched replied:
No because you can't animate display: none;

anthonydillon commented on jujugui/static/gui/src/app/components/user-profile/_user-profile.scss:
```diff
@@ -1,7 +1,67 @@
+            .deployment-input__field {
+                padding: 0;
+            }
+            .button--inline-neutral.first {
```
I think it would be better to add an additional contextual classes to piggy back this functional styling. 

In case someone changes neutral to base or positive we don't want this to break functionality.

hatched replied:
great idea

anthonydillon commented on jujugui/static/gui/src/app/components/user-profile/_user-profile.scss:
```diff
@@ -1,7 +1,67 @@
+        .button--inline-neutral {
+            overflow: hidden;
+            white-space: nowrap;
+            transition: all 0.5s ease-in;
```
This will override the transition on hover. Could to transition only the style you are changing?

hatched replied:
fixed

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5199/
Test PASSed.

hatched commented:
Thanks everyone! :shipit:

jujugui commented:
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge
